include/random_walks/billiard_shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,203 @@
+#ifndef RANDOM_WALKS_BILLIARD_SHAKE_AND_BAKE_WALK_HPP
+#define RANDOM_WALKS_BILLIARD_SHAKE_AND_BAKE_WALK_HPP
+
+#include <Eigen/Eigen>
+#include <cmath>
+#include <stdexcept>
+#include "sampling/sphere.hpp"
+#include "convex_bodies/hpolytope.h"
+#include "convex_bodies/convex_body.h"
+#include "random_walks/compute_diameter.hpp"
+#include "random_walks/accelerated_billiard_walk_utils.hpp"
+
+struct BilliardShakeAndBakeWalk
+{
+    BilliardShakeAndBakeWalk(double L) : param(L, true) {}
+    BilliardShakeAndBakeWalk() : param(0, false) {}
+
+    struct parameters
+    {
+        parameters(double L, bool set) : m_L(L), set_L(set) {}
+        double m_L;  bool set_L;
+    };
+
+    parameters param;
+
+    template <typename Polytope, typename RandomNumberGenerator>
+    struct Walk
+    {
+        using Point    = typename Polytope::PointType;
+        using NT       = typename Point::FT;
+        using VT       = typename Polytope::VT;
+        using MT       = typename Polytope::MT;
+        using MTdense  = Eigen::Matrix<NT,Eigen::Dynamic,Eigen::Dynamic>;
+
+        struct update_parameters { int facet_prev{-1}; } params_;
+        static constexpr NT kDefaultEps = NT(1e-10);
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope &P, const Point &boundary_pt, int facet_idx,
+             RandomNumberGenerator &rng, NT eps = kDefaultEps)
+        : P_(P), epsilon_(eps)
+        {
+            if(!P_.is_normalized()) P_.normalize();
+            _Len = compute_diameter<GenericPolytope>::template compute<NT>(P_);
+            initialize(boundary_pt, facet_idx);
+        }
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope &P, const Point &boundary_pt, int facet_idx,
+             RandomNumberGenerator &rng, NT eps, parameters const& par)
+        : P_(P), epsilon_(eps)
+        {
+            if(!P_.is_normalized()) P_.normalize();
+            _Len = par.set_L ? par.m_L : compute_diameter<GenericPolytope>::template compute<NT>(P_);
+            initialize(boundary_pt, facet_idx);
+        }
+
+        NT get_epsilon() const noexcept { return epsilon_; }
+
+
+        void apply(unsigned walk_len, RandomNumberGenerator& rng)
```

Comment:
I have very similar comments as I had for the SB implementation.
You should follow the accelerated billiard walk implementation. I think the best way to go is the following:
1. use exactly the same boundary oracles, but also add a facet index as input to skip when computing which facet the ray hits.
2. Av, Ar should not be touched in apply function
3.  The initialize function can do what the initiallize function does in SB